File: py/piano.py
import csv
import requests
from decouple import config

# Variables for request
AID = config("AID")
API_TOKEN = config("API_TOKEN")
URL = f"https://sandbox.piano.io/api/v3/publisher/user/list?api_token={API_TOKEN}&aid={AID}"

# Http Request
request = requests.get(url=URL)
request = request.json()
api_emails = list()
api_users = list()

for user in request["users"]:
    api_emails.append(user["email"])
    api_users.append(user)

# Save files into variables and read values
file_a = csv.reader(open("A.csv"))
file_b = csv.reader(open("B.csv"))

usersA = list()
usersB = list()

for row in file_a:
    user = {"user_id": row[0], "email": row[1]}
    usersA.append(user)
usersA.pop(0)

for row in file_b:
    user = {
        "user_id": row[0],
        "first_name": row[1],
        "last_name": row[2],
    }
    usersB.append(user)
usersB.pop(0)

# Merge values into usersA
for user in usersB:

    try:
        index = [
            x["user_id"] for x in usersA
        ].index(user["user_id"])
    except ValueError:
        index = -1

    usersA[index]["first_name"] = user[
        "first_name"
    ]
    usersA[index]["last_name"] = user["last_name"]

# Create output data variable, filtering if a user is already in database
output_data = list()
for user in usersA:
    output_data.append(list(user.values()))


for user in output_data:
    if user[1] in api_emails:
        try:
            index = [
                x["email"] for x in api_users
            ].index(user[1])
        except ValueError:
            index = -1
        if index != -1:
            user[0] = api_users[index]["uid"]


# Write the output file
output_filename = "C.csv"
with open(
    output_filename, "w", newline=""
) as file:
    csvwriter = csv.writer(file)
    csvwriter.writerow(
        [
            "user_id",
            "email",
            "first_name",
            "last_name",
        ]
    )
    csvwriter.writerows(output_data)
    print("File successfully created")

File: js/piano.js
const fs = require('fs');
const csv = require('fast-csv');
const axios = require('axios');
const { on } = require('events');
require('dotenv').config();

// Set of variables for the procedure
const file1 = 'A.csv',
	file2 = 'B.csv';
const AID = process.env.AID,
	API_TOKEN = process.env.API_TOKEN;
const URL = `https://sandbox.piano.io/api/v3/publisher/user/list?api_token=${API_TOKEN}&aid=${AID}`;
const usersA = [],
	usersB = [];
let apiUsers = [];
let apiEmails = [];

// Declare each async process as a promise

const requestPromise = new Promise((resolve) => {
	axios.get(URL).then((request) => {
		apiUsers = request.data.users;
		apiEmails = request.data.users.map((user) => user.email);
		console.log('USERS', apiUsers.length, typeof apiUsers);
		resolve();
	});
});

const file1Promise = new Promise((resolve) => {
	csv.parseFile(file1, { headers: true })
		.on('data', (data) => {
			const user = {
				user_id: data.user_id,
				email: data.email,
			};
			usersA.push(user);
		})
		.on('end', () => {
			console.log('finish A');
			resolve();
		});
});

const file2Promise = new Promise((resolve) => {
	csv.parseFile(file2, { headers: true })
		.on('data', (data) => {
			const user = {
				user_id: data.user_id,
				first_name: data.first_name,
				last_name: data.last_name,
			};
			usersB.push(user);
		})
		.on('end', () => {
			console.log('finish B');
			resolve();
		});
});

//Resolve all async processes concurrently
Promise.all([file1Promise, file2Promise, requestPromise]).then(() => {
	usersB.forEach((user) => {
		let index = usersA.findIndex((userA) => userA.user_id === user.user_id);
		usersA[index].first_name = user.first_name;
		usersA[index].last_name = user.last_name;
	});

	// Filter users if repeated in API
	usersA.forEach((user) => {
		let index = apiUsers.findIndex(
			(apiUser) => apiUser.email === user.email
		);
		if (index !== -1) {
			let correctId = apiUsers[index].uid;
			user.user_id = correctId;
		}
	});
	console.table(usersA);

	// Write the output file

	const csvStream = csv.format({
		headers: true,
	});
	const writableStream = fs.createWriteStream('C.csv');
	writableStream.on('finish', () => {
		console.log('DONE');
	});

	csvStream.pipe(writableStream);
	usersA.forEach((data) => csvStream.write(data));
	csvStream.end();
});
